compiler/src/main/java/com/squareup/anvil/compiler/codegen/CodeGenerationExtension.kt

```diff
@@ -227,47 +223,56 @@ internal class CodeGenerationExtension(
       }
     }
 
-    fun Collection<CodeGenerator>.generateCode(files: Collection<KtFile>): List<KtFile> =
+    fun Collection<CodeGenerator>.generateAndCache(files: Collection<KtFile>): List<KtFile> =
       flatMap { codeGenerator ->
         codeGenerator.generateCode(generatedDir, anvilModule, files)
+          .onEach { file ->
+            onGenerated(
+              generatedFile = file,
+              codeGenerator = codeGenerator,
+              allowOverwrites = false,
+            )
+          }
+          .toKtFiles(psiManager, anvilModule)
+      }
+
+    fun Collection<FlushingCodeGenerator>.flush(): List<KtFile> =
+      flatMap { codeGenerator ->
+        codeGenerator.flush(generatedDir, anvilModule)
           .onEach {
             onGenerated(
               generatedFile = it,
               codeGenerator = codeGenerator,
-              allowOverwrites = false,
+              // flushing code generators write the files but no content during normal rounds.
+              allowOverwrites = true,
             )
           }
           .toKtFiles(psiManager, anvilModule)
       }
 
-    fun Collection<CodeGenerator>.flush(): List<KtFile> =
-      filterIsInstance<FlushingCodeGenerator>()
-        .flatMap { codeGenerator ->
-          codeGenerator.flush(generatedDir, anvilModule)
-            .onEach {
-              onGenerated(
-                generatedFile = it,
-                codeGenerator = codeGenerator,
-                // flushing code generators write the files but no content during normal rounds.
-                allowOverwrites = true,
-              )
-            }
-            .toKtFiles(psiManager, anvilModule)
-        }
-
-    var newFiles = nonPrivateCodeGenerators.generateCode(files)
-
-    while (newFiles.isNotEmpty()) {
-      // Parse the KtFile for each generated file. Then feed the code generators with the new
-      // parsed files until no new files are generated.
-      newFiles = nonPrivateCodeGenerators.generateCode(newFiles)
+    fun List<CodeGenerator>.loopGeneration() {
+      var newFiles = generateAndCache(anvilModule.allFiles.toList())
```

Comment:
Previously `generateAndCache` (which was called `generateCode` before) was called once with `files` (like 258 of the removed code `var newFiles = nonPrivateCodeGenerators.generateCode(files)`). Always using `anvilModule.allFiles.toList()` can break custom code generators that use [topLevelFunctionReferences](https://github.com/square/anvil/blob/main/compiler-utils/src/main/java/com/squareup/anvil/compiler/internal/reference/AnvilModuleDescriptor.kt#L73-L80) or [topLevelPropertyReferences](https://github.com/square/anvil/blob/main/compiler-utils/src/main/java/com/squareup/anvil/compiler/internal/reference/AnvilModuleDescriptor.kt#L82-L89) because it won't contain any `KtFile` that doesn't contain a `KtClassOrObject`. So a `KtFile` that only contains top level functions and/or properties will not be passed to code generators anymore and because of that `topLevel*References` won't include functions/properties from those files.